File: backend/app/embeddings.py
import math
from typing import List


def embed_text(text: str, dim: int = 64) -> List[float]:
  """Deterministic stub embedding; replace with real model call."""
  vec = [0.0] * dim
  for i, ch in enumerate(text):
    vec[i % dim] += (ord(ch) * (i + 1)) % 97
  norm = math.sqrt(sum(v * v for v in vec)) or 1.0
  return [round(v / norm, 6) for v in vec]


def cosine_similarity(a: List[float], b: List[float]) -> float:
  dot = sum(x * y for x, y in zip(a, b))
  norm_a = math.sqrt(sum(x * x for x in a)) or 1.0
  norm_b = math.sqrt(sum(y * y for y in b)) or 1.0
  return dot / (norm_a * norm_b)

File: frontend/src/lib/embeddings.ts
import { chunkText } from "./utils";

// Placeholder embedding generator. Swap with real model call (OpenAI, Voyage, etc.).
export function embedText(text: string): number[] {
  const parts = chunkText(text, 16);
  const vector = new Array(32).fill(0);
  parts.forEach((part, idx) => {
    const hash = Array.from(part).reduce((sum, ch, i) => sum + ch.charCodeAt(0) * (i + 1), 0);
    vector[idx % vector.length] += hash % 997;
  });
  const norm = Math.sqrt(vector.reduce((s, v) => s + v * v, 0)) || 1;
  return vector.map((v) => v / norm);
}
